packages/interactive_media_ads/android/src/main/kotlin/dev/flutter/packages/interactive_media_ads/VideoAdPlayerProxyApi.kt

```diff
@@ -0,0 +1,67 @@
+// Copyright 2013 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+package dev.flutter.packages.interactive_media_ads
+
+import com.google.ads.interactivemedia.v3.api.AdPodInfo
+import com.google.ads.interactivemedia.v3.api.player.AdMediaInfo
+import com.google.ads.interactivemedia.v3.api.player.VideoAdPlayer
+import com.google.ads.interactivemedia.v3.api.player.VideoProgressUpdate
+
+class VideoAdPlayerProxyApi(override val pigeonRegistrar: ProxyApiRegistrar) :
+    PigeonApiVideoAdPlayer(pigeonRegistrar) {
+  override fun pigeon_defaultConstructor(): VideoAdPlayer {
+    return VideoAdPlayerImpl(this)
+  }
+
+  internal class VideoAdPlayerImpl(val api: VideoAdPlayerProxyApi) : VideoAdPlayer {
+    var savedVolume: Int = 0
+
+    var savedAdProgress: VideoProgressUpdate = VideoProgressUpdate.VIDEO_TIME_NOT_READY
+
+    override fun getAdProgress(): VideoProgressUpdate {
+      return savedAdProgress
+    }
+
+    override fun getVolume(): Int {
+      return savedVolume
+    }
+
+    override fun addCallback(callback: VideoAdPlayer.VideoAdPlayerCallback) {
+      api.pigeonRegistrar.runOnMainThread { api.addCallback(this, callbackArg = callback) {} }
+    }
+
+    override fun loadAd(adMediaInfo: AdMediaInfo, adPodInfo: AdPodInfo) {
+      api.pigeonRegistrar.runOnMainThread { api.loadAd(this, adMediaInfo, adPodInfo) {} }
+    }
+
+    override fun pauseAd(adMediaInfo: AdMediaInfo) {
+      api.pigeonRegistrar.runOnMainThread { api.pauseAd(this, adMediaInfo) {} }
+    }
+
+    override fun playAd(adMediaInfo: AdMediaInfo) {
+      api.pigeonRegistrar.runOnMainThread { api.playAd(this, adMediaInfo) {} }
+    }
+
+    override fun release() {
+      api.pigeonRegistrar.runOnMainThread { api.release(this) {} }
+    }
+
+    override fun removeCallback(callback: VideoAdPlayer.VideoAdPlayerCallback) {
+      api.pigeonRegistrar.runOnMainThread { api.removeCallback(this, callbackArg = callback) {} }
+    }
+
+    override fun stopAd(adMediaInfo: AdMediaInfo) {
+      api.pigeonRegistrar.runOnMainThread { api.stopAd(this, adMediaInfo) {} }
+    }
+  }
+
+  override fun setVolume(pigeon_instance: VideoAdPlayer, value: Long) {
+    (pigeon_instance as VideoAdPlayerImpl).savedVolume = value.toInt()
+  }
+
+  override fun setAdProgress(pigeon_instance: VideoAdPlayer, progress: VideoProgressUpdate) {
+    (pigeon_instance as VideoAdPlayerImpl).savedAdProgress = progress
+  }
```

Comment:
Custom methods added since `VideoAdPlayer.getVolume` and `VideoAdPlayer.getAdProgress` require a synchronous callback method.

Reply:
I think it's worth adding this as a code comment to these methods. It would be good to have it clarify why it's not calling `api.foo` in addition to saving the value.